### PR TITLE
Further reduce number of scenarios where we force rebuild the flat tree

### PR DESCRIPTION
**Pull request by nirvn**
This PR further optimizes the flat tree model by avoiding (as much as possible) model resets when rows are inserted and removed from the source model.

I was hoping we could avoid adding this level of complexity, but it turns out force rebuilding the proxy model on all row insertion deletion creates bad speed regressions.

For e.g., @RatInTheField reported a regression whereas adding a feature freezes QField for ~15 seconds. The reason why that is is that the layers he's using to add new features have [x] show feature count enabled, which in turn triggers _three_ pair of rows removed / inserted (the layer tree removes and re-add the layer's legend children to repopulate the count).

So, here we are.

Handling insertion of rows isn't too complex.  But for rows being deleted, it's slightly more complex. And in some instances, we simply have to reset the model as the stored QModelIndexes get out of sync. Eg.:

```
A
- 1
- 2
B
- 1
C
- 1
- 2
-- i
-- ii
- 3
```

If we remove A's children 1 and 2, that's easy enough, we merely have to adjust the flattened row count.

If we remove A's child 1, it's a bit more complex as we need to re-create the QModelIndex for A's child 2 (since it has moved from row() = 1 to row()  = 0. We can still handle this.

However, it we are to remove C's child 1, we enter in a scenario that becomes too complex to handle. We would not only have to re-create the QModelIndex for C's children 2 and 3, we would have to re-do the children of those too, etc. etc. In this scenario, we revert back to resetting the model.

With this in place, addition of features into a vector layer that has [x] show feature count is now amazingly fast. It also fixes a project load regression when projects include layer(s) with [x] show feature count enabled.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1418-0fb73df10563b2edbccede18ca6d23c34755e862-arm64_v8a.apk)

Other architectures: [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1418-0fb73df10563b2edbccede18ca6d23c34755e862-armv7.apk), [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1418-0fb73df10563b2edbccede18ca6d23c34755e862-x86_64.apk), [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1418-0fb73df10563b2edbccede18ca6d23c34755e862-x86.apk)

File: src/core/layertreemodel.cpp [hunk 1]
```diff
@@ -62,6 +62,170 @@ void FlatLayerTreeModel::updateMap( const QModelIndex &topLeft, const QModelInde
...
 
+void FlatLayerTreeModel::insertInMap( const QModelIndex &parent, int first, int last )
+{
+  if ( mFrozen )
+    return;
+
+  int insertedAt = -1;
+  if ( first == 0 )
+  {
+    if ( !parent.isValid() )
```
Comment (m-kuhn) [inline src/core/layertreemodel.cpp:73]: Do we need to consider the row as well?

File: src/core/layertreemodel.cpp [hunk 2]
```diff
@@ -62,6 +62,170 @@ void FlatLayerTreeModel::updateMap( const QModelIndex &topLeft, const QModelInde
...
+    }
+  }
+
+  if ( insertedAt >  -1 )
+  {
+    beginInsertRows( QModelIndex(), insertedAt, insertedAt + ( last - first ) );
+
+    QMap<int, int> treeLevelMap;
+    mIndexMap.clear();
+    for ( const auto &index : mRowMap.keys() )
```
Comment (m-kuhn) [inline src/core/layertreemodel.cpp:97]:
```suggestion
    const auto keys = mRowMap.keys();
    for ( const auto &index : keys )
```

File: src/core/layertreemodel.cpp [hunk 3]
```diff
@@ -62,6 +62,170 @@ void FlatLayerTreeModel::updateMap( const QModelIndex &topLeft, const QModelInde
...
+        resetNeeded = true;
+        break;
+      }
+      modifiedUntil++;
+    }
+    modifiedUntil--;
+
+    if ( resetNeeded )
+    {
+      // Removed rows can't be handled, modle reset needed
```
Comment (m-kuhn) [inline src/core/layertreemodel.cpp:180]:
```suggestion
      // Removed rows can't be handled, model reset needed
```

**Review by m-kuhn — APPROVED** [review]
Wow, this thing becomes :unicorn:

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available:
  - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1418-302214ce0efcbf5c96a5eaca2e8e5cffa5d929d0-arm64_v8a.apk)

Other architectures: [armv7](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1418-302214ce0efcbf5c96a5eaca2e8e5cffa5d929d0-armv7.apk), [x86_64](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1418-302214ce0efcbf5c96a5eaca2e8e5cffa5d929d0-x86_64.apk), [x86](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1418-302214ce0efcbf5c96a5eaca2e8e5cffa5d929d0-x86.apk)

**Comment by nirvn** [comment]
WIP, there's nothing WIP here!

**Comment by m-kuhn** [comment]
RIP WIP: you've been replaced by draft PRs

**Comment by nirvn** [comment]
@m-kuhn , thanks :)

**Pull request merged**